### ISSUE TITLE
Non-Reproducible Results in LDDMM While Aligning Xenium Datasets

### ISSUE THREAD

**Issue by qu-bit1** [JEFworks-Lab/STalign#45]
Hey!

While aligning two Xenium datasets (V1 and Prime 5K) from [10x Genomics](https://www.10xgenomics.com/datasets/xenium-human-lung-cancer-post-xenium-technote), we encountered non-reproducible results when running LDDMM with the same parameters. This issue persists across different hardware, including CPU and A40 GPU.

Observations:
- **Inconsistent Alignment Results**: The initial run produced well-aligned images, but subsequent runs with the same parameters failed to do so.

First successful run:
![Image](https://github.com/user-attachments/assets/69106d18-579c-47d6-9eef-fca3123fa481)

Random run with identical parameters:
![Image](https://github.com/user-attachments/assets/8a45a0d0-5cdf-4e03-a5e3-7d89df1c65e3)

- **Non-Converging Runs**: In some cases, the LDDMM process runs indefinitely without converging.

You can check out our implementation here: [STalign.ipynb](https://github.com/DebDDash/UGP/blob/master/week4_alignment/STalign/STalign.ipynb).

Any insights into potential causes or fixes would be greatly appreciated. 
Thanks!

**Comment by JEFworks** [comment]
Dear Sagar,

Thanks for reaching out and for your interest in STalign. Thanks for the useful notebook and well documented examples. 

As you may recall from the original publication (https://www.nature.com/articles/s41467-023-43915-7), STalign uses a gradient descent that is stochastic in nature. This means, every run will be a little different; you will need to set a random seed if you want to fully reproduce results. 

You are correct that not every run of LDDMM will converge to a stable global minimum. This is because the gradient descent may get stuck in a local minima. This is why it is important to use the diagnostic plots as you are doing. 

Looking at your diagnostic plots, you may wish to use larger step sizes or use some anchoring points to initialize the affine component of your alignment for better convergence. This tutorial may be useful: https://jef.works/STalign/notebooks/xenium-xenium-alignment.html

Your weights also look a bit off; instead of highlighting just the tissue, the background is also being prioritized in the alignment. You may wish to change the following hyperparameters: 'sigmaM', 'sigmaB', and 'sigmaA' . This tutorial may be useful: https://jef.works/STalign/notebooks/merfish-allen3Datlas-alignment.html

Hope that helps,
Prof. Fan
